crates/ntx-builder/src/state.rs

```diff
@@ -0,0 +1,196 @@
+                    },
+                });
+
+                self.inflight.insert(
+                    id,
+                    InflightTx {
+                        account_delta: account_id,
+                        nullifiers: nullifiers.clone(),
+                        notes: network_notes.iter().map(NetworkNote::nullifier).collect(),
+                    },
+                );
+
+                self.notes
+                    .extend(network_notes.into_iter().map(|note| (note.nullifier(), note)));
+                self.nullifiers.extend(nullifiers);
+            },
+            MempoolEvent::BlockCommitted { header, txs } => {
+                assert!(header.prev_block_commitment() == self.latest_header.commitment());
+
+                self.latest_header = header;
+
+                for tx in txs {
+                    let Some(tx) = self.inflight.remove(&tx) else {
+                        continue;
+                    };
+
+                    if let Some(account_id) = tx.account_delta {
+                        self.accounts
+                            .get_mut(&account_id)
+                            .expect("account with delta should be tracked")
+                            .commit_one();
+                    };
+
+                    for nullifier in tx.nullifiers {
+                        self.notes.remove(&nullifier);
+                        self.nullifiers.remove(&nullifier);
+                    }
+                }
+            },
+            MempoolEvent::TransactionsReverted(txs) => {
+                for tx in txs {
+                    let Some(tx) = self.inflight.remove(&tx) else {
+                        continue;
+                    };
+
+                    if let Some(account_id) = tx.account_delta {
+                        let account = self
+                            .accounts
+                            .remove(&account_id)
+                            .expect("account with delta should be tracked")
+                            .revert_one();
+
+                        if let Some(account) = account {
+                            self.accounts.insert(account_id, account);
+                        }
+
+                        for nullifier in tx.nullifiers {
+                            self.nullifiers.remove(&nullifier);
+                        }
+
+                        for note in tx.notes {
+                            self.notes.remove(&note);
+                        }
+                    };
+                }
+            },
+        }
+    }
+
+    pub fn chain_tip(&self) -> BlockNumber {
+        self.latest_header.block_num()
+    }
+
+    pub fn account_with_unconsumed_notes(&self) -> Option<(&Account, Iter<&NetworkNote>)> {
+        todo!()
+    }
+}
+
+struct InflightTx {
+    account_delta: Option<AccountId>,
+    nullifiers: Vec<Nullifier>,
+    notes: Vec<Nullifier>,
+}
+
+struct AccountState {
+    state: Account,
+    deltas: VecDeque<AccountDelta>,
+}
```

Comment:
Would be great to add some comments here. My understanding is that `state` is the state of the account before any of the `deltas` have been applied to it, right? Also, we are not guaranteed to have at least one delta (i.e., `deltas` could be empty).